src/server/gc_worker/applied_lock_collector.rs

```diff
@@ -0,0 +1,732 @@
+                    return;
+                }
+            };
+
+            if lock.ts <= max_ts {
+                let key = Key::from_encoded_slice(put_request.get_key());
+                self.send(LockObserverMsg::Locks(vec![(key, lock)]));
+            }
+        }
+    }
+}
+
+impl ApplySnapshotObserver for LockObserver {
+    fn pre_apply_plain_kvs(
+        &self,
+        _: &mut ObserverContext<'_>,
+        cf: CfName,
+        kv_pairs: &[(Vec<u8>, Vec<u8>)],
+    ) {
+        if cf != CF_LOCK {
+            return;
+        }
+
+        let max_ts = self.load_max_ts();
+        if max_ts.is_zero() {
+            return;
+        }
+
+        let locks: Result<Vec<_>> = kv_pairs
+            .iter()
+            .map(|(key, value)| {
+                Lock::parse(value)
+                    .map(|lock| (key, lock))
+                    .map_err(|e| ErrorInner::Mvcc(e.into()).into())
+            })
+            .filter(|result| result.is_err() || result.as_ref().unwrap().1.ts <= max_ts)
+            .map(|result| {
+                // `pre_apply_plain_keys` will be invoked with the data_key in RocksDB layer. So we
+                // need to remove the `z` prefix.
+                result.map(|(key, lock)| (Key::from_encoded_slice(origin_key(key)), lock))
+            })
+            .collect();
+
+        match locks {
+            Err(e) => self.send(LockObserverMsg::Err(e)),
+            Ok(l) => self.send(LockObserverMsg::Locks(l)),
+        }
+    }
+
+    fn pre_apply_sst(&self, _: &mut ObserverContext<'_>, cf: CfName, _path: &str) {
+        if cf == CF_LOCK {
+            let e = box_err!("snapshot of lock cf applied from sst file");
+            error!("cannot collect all applied lock"; "err" => ?e);
+            self.send(LockObserverMsg::Err(e));
+        }
+    }
+}
+
+struct LockCollectorRunner {
+    observer_state: Arc<LockObserverState>,
+
+    collected_locks: Vec<(Key, Lock)>,
+
+    /// `is_clean` is true, only it's sure that all applying of stale locks (locks with start_ts <=
+    /// specified max_ts) are monitored and collected. If there are too many stale locks or any
+    /// error happens, `is_clean` must be set to `false`.
+    /// TODO: when dirty, stop collecting.
+    is_clean: bool,
+}
+
+impl LockCollectorRunner {
+    pub fn new(observer_state: Arc<LockObserverState>) -> Self {
+        Self {
+            observer_state,
+            collected_locks: vec![],
+            is_clean: true,
+        }
+    }
+
+    fn handle_observer_msg(&mut self, msg: LockObserverMsg) {
+        if !self.is_clean {
+            return;
+        }
+
+        match msg {
+            LockObserverMsg::Err(e) => {
+                self.is_clean = false;
+                info!("lock collector marked dirty because received error"; "err" => ?e);
+            }
+            LockObserverMsg::Locks(mut locks) => {
+                if locks.len() + self.collected_locks.len() > MAX_COLLECT_SIZE {
```

Comment:
Do we need to continue collecting locks when it is already dirty?

Reply:
I think it's ok to collect it, so that TiDB can resolve them before next retry or falling back.

Reply:
According to the current implementation, it seems that the subsequently collected locks are just dropped because of the truncation.

Reply:
Oh, yes, I didn't collect them in the current code. I'll change it.